demos/guide/quic-hq-interop.c

```diff
@@ -0,0 +1,1124 @@
+     * mode in this demo.
+     */
+    if (!SSL_set_blocking_mode(*ssl, 0)) {
+        fprintf(stderr, "Failed to turn off blocking mode\n");
+        goto end;
+    }
+
+    /* Do the handshake with the server */
+    while ((ret = SSL_connect(*ssl)) != 1) {
+        if (handle_io_failure(*ssl, ret) == 1)
+            continue; /* Retry */
+        fprintf(stderr, "Failed to connect to server\n");
+        goto end; /* Cannot retry: error */
+    }
+
+    return 1;
+end:
+    SSL_CTX_free(*ctx);
+    SSL_free(*ssl);
+    BIO_ADDR_free(peer_addr);
+    return 0;
+}
+
+/**
+ * @brief Entry point for the QUIC hq-interop client demo application.
+ *
+ * This function sets up an SSL/TLS connection using QUIC, sends HTTP GET
+ * requests for files specified in the command-line arguments, and saves
+ * the responses to disk. It handles various configurations such as IPv6
+ * support, session caching, and key logging.
+ *
+ * @param argc The number of command-line arguments.
+ * @param argv The array of command-line arguments. The expected format is
+ *             "[-6] hostname port file".
+ * @return EXIT_SUCCESS on success, or EXIT_FAILURE on error.
+ *
+ * @note The function performs the following main tasks:
+ *       - Parses command-line arguments and configures IPv6 if specified.
+ *       - Reads the list of requests from the specified file.
+ *       - Sets up the SSL context and configures certificate verification.
+ *       - Optionally enables key logging and session caching.
+ *       - Establishes a non-blocking QUIC connection to the server.
+ *       - Sends an HTTP GET request for each file and writes the response
+ *         to the corresponding output file.
+ *       - Gracefully shuts down the SSL connection and frees resources.
+ *       - Prints any OpenSSL error stack information on failure.
+ */
+int main(int argc, char *argv[])
+{
+    SSL_CTX *ctx = NULL;
+    SSL *ssl = NULL;
+    BIO *req_bio = NULL;
+    int res = EXIT_FAILURE;
+    int ret;
+    char req_string[1024];
+    size_t readbytes = 0;
+    char buf[160];
+    int eof = 0;
+    int argnext = 1;
+    char *reqfile = NULL;
+    char *reqnames = OPENSSL_zalloc(1025);
+    size_t read_offset = 0;
+    size_t bytes_read = 0;
+    size_t poll_idx = 0;
+    size_t poll_done = 0;
+    size_t result_count = 0;
+    struct timeval poll_timeout;
+    size_t this_poll_count = 0;
+    char *req = NULL;
+    char *hostname, *port;
+    int ipv6 = 0;
+
+    if (argc < 4) {
+        fprintf(stderr, "Usage: quic-hq-interop [-6] hostname port file\n");
+        goto end;
+    }
+
+    if (!strcmp(argv[argnext], "-6")) {
+        if (argc < 5) {
+            fprintf(stderr, "Usage: quic-hq-interop [-6] hostname port\n");
+            goto end;
+        }
+        ipv6 = 1;
+        argnext++;
+    }
+    hostname = argv[argnext++];
+    port = argv[argnext++];
+    reqfile = argv[argnext];
+
+    memset(req_string, 0, 1024);
```

Comment:
I'm good with the use of a macro, I'm less sure about the elimination of memset, as I don't think C mandates that ELF .bss sections are mandates as initialized to 0 (or am I wrong about that)?

Reply:
You're wrong. Static data is always initialized to 0.

Reply:
yeah, I looked it up, and its required, though this can be simplified more anyway.  req_string is actually unused in main (holdover from previous development).  the only reason it doesn't warn during compilation is because of that memset, so both can just be removed.  The new location for req_string is in build_request_set(), where we use BIO_snprintf which will write a null terminator to the proper location